Operator SDK supports OLM deleting operands (#63)
operator-framework/enhancements: enhancements/sdk-supports-operands-deletion.md

---
title: sdk-supports-operands-deletion
authors:
 - "@rashmigottipati"
reviewers:
  - "@ecordell"
  - "@estroz"
  - "@jmrordi"
  - "@joelanford"
approvers:
  - "@ecordell"
  - "@estroz"
  - "@jmrordi"
  - "@joelanford"
creation-date: 2021-01-29
last-updated: 2021-01-29
status: provisional
---

# sdk-supports-operands-deletion

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Summary

> The purpose of this enhancement is to outline how the [Operator SDK][sdk] can support operator developers targeting [Operator Lifecycle Manager (OLM)][olm] and need to delete operands. SDK will provide mechanisms to enable and faciliate operand deletion for operators that are installed using OLM and operator developers can explicitly opt-in for the cleanup of their operator workloads.

Reviewer: CR deletion involves operand cleanup, but they are not equivalent
```suggestion
The purpose of this enhancement is to outline how the [Operator SDK (OSDK)][sdk] can support operator developers targeting [Operator Lifecycle Manager (OLM)][olm] and need to clean up operands prior to an uninstall via custom resource (CR) deletion. OSDK will provide mechanisms to enable and facilitate CR deletion for operators that are installed using OLM and operator developers can explicitly opt-in for the cleanup of their operator workloads.
```

For this cleanup proposal’s perspective the Custom Resource (CR) object managed by the operator is the only resource object that is considered for cleanup. Other resources such as pods, volumes and off-cluster resources created by the operator are also operands but those resources will get auto-cleaned as a result of removing the CRs.

In kubernetes garbage collection model, some objects are owners of other objects based on the resource hierarchy. Dependent objects have [OwnerReference][owner-reference] field that points to the owning object. This ownership can also be specified manually by the user. If user deletes the owner object, Kubernetes garbage collector will delete all of its dependent objects (sometimes automatically if it is specified).

However, cross-namespaced owner references are not allowed by design and this is where the operator would need to leverage using [Finalizers][finalizers]. to perform cleanup of operands by processing certain deletion events.

SDK will encourage and enable the use of finalizers when it is required to. By default, all operators would have scaffolded logic for Finalizers.

> 

Author: Are you referring to the cleanup command specifically, or OSDK's approach in handling operand cleanup?

## Goals

- Provide default scaffolding for an operand cleanup
- Cleanup is strictly on an opt-in basis only and not automatic by default
- Cleanup is limited to removal of CRs and not CRDs
- SDK will have additional flags to prevent deletion of CRDs and instead use OLM's operand cleanup feature

## Non-Goals

- SDK support in handling cleaning up resources other than CRs  

## Possible Future Goals (Not in scope)

- SDK should have a mechanism to handle events/alerts received by OLM in case cleanup is blocked
- The timeout for operand deletion should be user configurable by the SDK

## Proposal

### User Stories

#### Story 1

As an operator author, I want an upstream finalizer no-op handler scaffolded by default and/or library to help write finalizers.

#### Story 2

As an operator author, I want scaffolded fields in the CSV generation by default to enable operand cleanup.

#### Story 3

As an operator author, I would like the Operator SDK's cleanup command to handle deletion of CRDs appropriately based on the flags provided.

### Implementation Details/Notes/Constraints

#### Upstream finalizer no-op handler scaffolded by default and/or library to help write Finalizers

> Any operator that handles delete events or if the operator is expecting a delete event for CRs should have Finalizers. The only time Finalizers may not be required is when OwnerReferences exist and they handle all of the cleanup. But as we tend towards cluster scoped operators, OwnerReferences won't be allowed. In that case, all operators should have Finalizers.

Reviewer: ```suggestion
Any operator that deletes or expects delete events for CRs that create resources can use finalizers or owner references, although the latter can be used when those CRs only create resources in the same namespace. Additionally finalizers permit customizable deletion steps, ex. update a CR's status during deletion with a particular message. Because finalizers are broadly applicable and give operator authors control over how CR deletion proceeds, operator projects would benefit from some initial finalizer code scaffolded by default.
```

> To enable this, it would be ideal to have a Finalizer no-op handler scaffolded by default in upstream such as Kubebuilder. The idea is to always scaffold out a Finalizer, a Delete Event Handler and a Reconciler that checks for Finalizers and has a no-op in cases where it's not required to use a Finalizer. 

Reviewer: ```suggestion
To enable this, it would be ideal to have a finalizer "no-op" handler, i.e. some code that does not clean up a CR but has a `TODO(user)` comment, scaffolded by default in the kubebuilder project. The idea is to always scaffold out a finalizer, a delete event handler and a Reconciler that checks for Finalizers and has a no-op in cases where it's not required to use a Finalizer. 
```
What do you mean by
>has a no-op in cases where it's not required to use a Finalizer.

> The Reconciler checks for presence of a Finalizer in a CR object. If the CR object is marked for deletion, which is indicated by the timestamp that was set, the Reconciler runs the finilization logic and removes the finalizer. If the finalization logic fails for some reason, then the Reconciler keeps the Finalizer to retry during the next reconcilation. Typically, the finalization logic would entail necessary cleanup steps that the operator needs to perform before the CR can be deleetd.

Reviewer: Needs a link to the [finalizer example](https://v1-3-x.sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#handle-cleanup-on-deletion) in operator-sdk docs
```suggestion
The Reconciler checks for presence of a Finalizer in a CR object. If the CR object is marked for deletion, indicated by `metadata.deletionTimestamp` being set, the Reconciler runs the finilizer logic and removes the finalizer. If the finalization logic fails for some reason, then the Reconciler keeps the finalizer to retry during the next reconcilation. Typically, the finalizer logic would entail necessary cleanup steps that the operator needs to perform before the CR can be deletetd.
```

> Note: If a Finalizer doesn't exist on the CR and the object is marked for delete event, the Reconciler would add a Finalizer to the CR object as part of reconcilcation.

Reviewer: Why would the operator do this? Shouldn't the finalizer be added to all CRs prior to deletion, so that full deletion doesn't occur until the finalizer is removed?

#### Configuration of the cleanup API in SDK

> A [ClusterServiceVersion (CSV)][csv] represents a particular version of a running operator on a cluster and is an actual installation of the operator. It's more appropriate to configure operand cleanup API on the CSV than any other resource object.

Reviewer: I would just say that "The CSV will support the `spec.cleanup` field once [the operand cleanup EP](https://github.com/operator-framework/enhancements/blob/master/enhancements/olm-deletes-operands.md) is implemented". There's no need to explain the reasoning behind why, which is in that EP.

Configuring the cleanup API elsewhere, such as [Subscription][subscription] for example, will remove the flexibility for a user to opt-in for operand cleanup if the CSV was installed manually without a subscription.

> To enable configuration of operand deletion on the CSV, Operator SDK will scaffold the cleanup fields in the CSV spec and enable it to be false by default. The user can then explicitly patch the CSV and set the cleanup field to true if the user would like to opt-in for automatic deletion of operands using OLM. The scaffolding logic of cleanup fields in the CSV may reside in go plugin implementation.

Reviewer: The `operator generate kustomize manifests` command will add this field to CSVs, not plugin code. CSVs are language-agnostic
```suggestion
To enable configuration of operand deletion on the CSV, Operator SDK will generate CSVs with the `spec.cleanup` field set to `false` by default. The user can then explicitly set this field to true to opt-in to automatic deletion of CRs using OLM.
```

#### Existing cleanup command in SDK

> The existing cleanup command in SDK would delete the CRDs and all other resources during cleanup. Ideally, the cleanup command should have certain flags that prevent deletion of CRDs, for future parity. Currently, the cleanup command is only used for testing purposes and not intended for production environments. But in the future, when OLM supports cleanup for global operators, SDK cleanup mechanism would have the desired behavior with these changes.

Reviewer: ```suggestion
The existing `operator-sdk cleanup` command currently deletes the CRDs and all other resources owned by the operator, without an option to disable this behavior. Ideally, the cleanup command should have certain flags that prevent deletion of CRDs, for feature parity with the [kubectl operator plugin](https://github.com/operator-framework/kubectl-operator/).
```

This should be in the future goals section

>Currently, the cleanup command is only used for testing purposes and not intended for production environments. But in the future, when OLM supports cleanup for global operators, SDK cleanup mechanism would have the desired behavior with these changes

> For example, the cleanup command could have a boolean flag such as `--delete-crds` which will delete all owned CRDs and all CRs if set to true.

Reviewer: `--delete-crds` is the flag you're proposing and not an example
```suggestion
The `--delete-crds` boolean flag should be added to the `cleanup` command, and defaulted to `true` to preserve current behavior. When `--delete-crds=false`, the command will not clean up CRDs.
```

```

> operator-sdk cleanup <operatorPackageName> --delete-all --delete-crds

Reviewer: What does `--delete-all` do? I see no discussion of this flag.

Cross-namespace owner references are not allowed by design; instead, a [finalizer][finalizers] should be used to perform cleanup. Finalizers are more powerful than owner references because they permit a controller to define how an object is cleaned up.

In OLM's [operand cleanup proposal][operand-cleanup-proposal] to handle operand deletion, OLM would configure the cleanup API in the [ClusterServiceVersion (CSV)][csv] and deleting the CSV will be the trigger to initiate operand cleanup. The CSV spec would have an optional field `spec.cleanup` which lets the user configure the intent to cleanup operands before the CSV is deleted.

```YAML
apiVersion: operators.coreos.com/v1alpha1
kind: ClusterServiceVersion
...
spec:
  ...
  cleanup:
    enabled: false
```

Upon setting `spec.cleanup.enabled: true`, OLM will add `operatorframework.io/cleanup-apis` finalizer on the CSV to prevent operator removal before cleanup. See [operand cleanup proposal][operand-cleanup-proposal] for more details.

## Goals

- Provide default scaffolding for an operand cleanup
- Cleanup is not opted in by default to preserve current behavior

> - `operator-sdk cleanup` should optionally delete CRDs, and use OLM's operand cleanup feature

Author: That behavior is still intact. Except that we are providing an option to not delete CRDs as it would give us more control and when OLM has the cleanup feature implemented, there won't be any more changes required for the cleanup command.

## Possible Future Goals (Not in scope)

- SDK should have a mechanism to handle events/alerts received by OLM in case cleanup is blocked

> - The timeout for operand deletion should be user configurable by the SDK

Author: It's implementable but we may not do this in the first pass, just to avoid the complexity that comes with it. 

If it is user configurable, the user/operator author needs to be aware of how long the cleanup process would take and in some cases they may not know what value it would be.

SDK could enforce a generic default timeout for all operators but that would mean that for a long running cleanup process, we might prematurely conclude that it's stuck and try to unblock operand deletion and potentially leave things lingering. So for now, we will not implement timeouts, and it will be one of the risks.

- Currently, the cleanup command in OSDK is only used for testing purposes and not intended for production environments. In the future, when OLM supports cleanup for global operators, SDK cleanup mechanism would have the desired behavior with changes described in [existing cleanup command in SDK][existing-cleanup-command-in-sdk] section.

## Proposal

### User Stories

#### Story 1

As an operator author, I want an upstream finalizer handler scaffolded by default that lets me easily fill in operand-specific cleanup code.

#### Story 2

> As an operator author, I want `spec.cleanup: false` added to my CSV by default when I call `operator-sdk generate kustomize manifests` so I can easily turn on the cleanup feature if I want to.

Author: This is definitely an option. We can provide the option where the user can automatically opt-in and explicitly provide the cleanup flag and set it to true. 

If a user doesn't provide the flag, SDK will apply the default and set `spec.cleanup: false` and leave it to the user to manually patch the csv if they intend to opt-in later. 

I will include this possibility in the EP.

#### Story 3

As an operator author, I would like `operator-sdk cleanup` to handle deletion of CRDs appropriately based on the flags provided.

### Implementation Details/Notes/Constraints

#### Upstream finalizer no-op handler scaffolded by default and/or library to help write Finalizers

Any operator that deletes or expects delete events for CRs that create resources can use finalizers or owner references, although the latter can be used when those CRs only create resources in the same namespace. Additionally finalizers permit customizable deletion steps, ex. update a CR's status during deletion with a particular message. Because finalizers are broadly applicable and give operator authors control over how CR deletion proceeds, operator projects would benefit from some initial finalizer code scaffolded by default.

> To enable this, it would be ideal to have a finalizer "no-op" handler, i.e. some code that does not clean up a CR but has a `TODO(user)` comment, scaffolded by default in the kubebuilder project. The idea is to always scaffold out a finalizer, a delete event handler and a Reconciler that checks for Finalizers and has a no-op in cases where it's not required to use a Finalizer.

Author: Do you mean "remove the finalizer" if it's not implemented?

The Reconciler checks for presence of a Finalizer in a CR object. If the CR object is marked for deletion, indicated by `metadata.deletionTimestamp` being set, the Reconciler runs the finilizer logic and removes the finalizer. If the finalization logic fails for some reason, then the Reconciler keeps the finalizer to retry during the next reconcilation. Typically, the finalizer logic would entail necessary cleanup steps that the operator needs to perform before the CR can be deleted. See [finalizer example][finalizer-example] for more details.

#### Configuration of the cleanup API in SDK

A [ClusterServiceVersion (CSV)][csv] represents a particular version of a running operator on a cluster and is an actual installation of the operator. The CSV will support the `spec.cleanup` field once the [operand cleanup proposal][operand-cleanup-proposal] is implemented

> To enable configuration of operand deletion on the CSV, the `operator-sdk generate kustomize manifests` command will generate CSVs with the `spec.cleanup` field set to `false` by default. The user can then explicitly set this field to true to opt-in to automatic deletion of CRs using OLM.

Author: Updated the EP to include this possibility.

#### Existing cleanup command in SDK

The existing `operator-sdk cleanup` command currently deletes the CRDs and all other resources owned by the operator, without an option to disable this behavior. Ideally, the cleanup command should have certain flags that prevent deletion of CRDs, for feature parity with the [kubectl operator plugin](https://github.com/operator-framework/kubectl-operator/).

Existing cleanup command in OSDK:

```
operator-sdk cleanup <operatorPackageName>
```

Proposed cleanup command with flags:

```
operator-sdk cleanup <operatorPackageName> --delete-crds --delete-all
```

> `--delete-crds` boolean flag should be added to the `cleanup` command, and defaulted to `true` to preserve current behavior. When `--delete-crds=false`, the command will not clean up CRDs.

Author: I don't think it can be done with cobra. Rethinking the flags, I think `--delete-all` is kind of redundant because of today the cleanup command removes everything. I removed this flag from the EP and now only have `[--delete-crds]` flag which will be set to false by default to preserve the current behavior of the cleanup command. 

If it is set to true by the user, then OSDK will not remove any CRDs.

Additionally, `--delete-all` flag should also be added on the `cleanup` command that will delete all owned CRDs, CRs and operator groups.

> These flags can be set accordingly to prevent deletion of CRDs unless explicitly asked for by the user.

Author: With only flag `--delete-crds`, it won't behave too differently, as we will default to removing everything to preserve current behavior if the flag is not set. But still, it would be a breaking change.

### Risks and Mitigations

> - What if Finalizers cause to hang? How do we handle this before implementing timeouts, and how do we mitigate this risk.

Author: Noted this point in the EP.
